[Team-09][BE][테리&벅픽] 3주차 PR 1번째 - 이슈 필터 검색 API 개발,  코드 전반 리팩토링

mybloom opened the pull request:
안녕하세요 `프레디`!
`벅픽, 테리` 입니다. 3주차 1번째 PR 입니다.

코드 전반 리팩토링과 이슈 필터 검색 API 를 개발하였습니다. 

리뷰 덕분에 메서드의 적절한 위치에 대해서 스스로 판단해볼 수 있게 되었습니다.
부족한 면들이 있지만 최대한 빨리 PR 보냈습니다. 

감사합니다! 😄 

</br>

### ERD 및 API docs

<details>
<summary> ERD </summary>
    
![image](https://user-images.githubusercontent.com/91416897/174514679-dcae6a60-c8a8-4284-9160-c3575af76cfe.png)
    
</details>

<details>
<summary> API docs </summary>
    
https://documenter.getpostman.com/view/20452796/UzBiNobv
    
</details>

</br>

### 작업 내용

- ✅ 이슈 목록 검색API 개발
    
- ✅ 4차 리뷰 반영
    - 코드 전체적으로 메서드의 위치 변경. 
      - 기존에 IssueService에 있던 로직을 각 도메인에 맞는 곳으로 이동

mybloom committed:
refactor: 더보기 화면 url 변경
 (bugpigg/issue-tracker#45)

 - issues/{id}/editing 에서 issues/{id}/moreInfo로 변경
 - url 네이밍이 수정할 것 같은 느낌이라 "추가정보"의 의미를 담는 더보기 이므로 변경

mybloom committed:
refactor: Milestone 객체를 생성 작업을 IssueService에서 MilestoneService 로 변경
 (bugpigg/issue-tracker#45)

 -  Milestone을 관리하는 곳에서 Milestone을 생성

mybloom committed:
refactor: contorller에서 requestBody의 유효성 체크하도록 수정. 기존에는 service에서 판별
 (bugpigg/issue-tracker#45)

 - 이슈 작성자가 작성한 마일스톤/라벨인지 확인
 - AssigneeId는 회원db에 등록된 값인지 확인
 - TODO : INSERT, UPDATE시 DB상에서 외래키로 판별된다면 작성자가 작성하고, 회원DB에 있는 값인지 확인할 필요가 없어보이긴 한다.(일단 두고 시간이 된다면 생각해보자)

mybloom committed:
refactor: 이슈 수정할 때는 할당자 변경하는 부분이 없어 assigneeIds 속성 제거
 (bugpigg/issue-tracker#45)

mybloom committed:
feat: 마일스톤id의 유효성 확인로직, Milestone 객체를 생성하는 task를 MilestoneService에 작성
 (bugpigg/issue-tracker#45)

mybloom committed:
feat: 라벨id의 유효성 확인로직 LabelService에 작성
 (bugpigg/issue-tracker#45)

mybloom committed:
feat: AssigneeId의 유효성 확인로직 IssueAssigneeService에 작성
 (bugpigg/issue-tracker#45)

mybloom committed:
refactor: 이슈에서 사용하는 유효성 로직을 모아두었던 IssueValidateService.java를 각 도메인 Service로 이동하여 삭제
 (bugpigg/issue-tracker#45)

mybloom committed:
refactor: 이슈 생성/수정 시 request값들의 유효성 로직을 Controller로 옮기고, 불필요한 local변수 삭제
 (bugpigg/issue-tracker#45)

mybloom committed:
refactor: Issue가 가진 issueLabels, issueAssignees의 컬렉션을 List에서 Set으로 변경
 (bugpigg/issue-tracker#45)

 - List일 경우 내부적으로 Bag을 사용하는데, Bag은 컬렉션의 내용이 변경될 경우 전체 삭제후 생성하는 작업을 한다고 하여 Set으로 변경

bugpigg committed:
[BE] 이슈 목록 조회 API 구현  (#47)

* fix: 검색 쿼리 작성을 위한 일대다 필드 추가

* feat: 검색 기능 추가(bugpigg/issue-tracker#46)

- 이슈 제목, 상태, 글쓴이, 할당자, 라벨, 마일스톤 조건을 기반으로 이슈목록을 조회하는 기능을 추가하였습니다.

mybloom committed:
refactor: 이슈 연관 엔티티 save 로직을 각각의 Service에서 수행하도록 수정
 (bugpigg/issue-tracker#45)

mybloom committed:
refactor: Issue에 각각의 setter를 없애고 update() 하나로 모두 수정할 수 있도록 변경
 (bugpigg/issue-tracker#45)

 - 메서드가 하는 일을 이름으로 파악할 수 있도록 하기 위해 수정

mybloom committed:
chore: 불필요한 import문 삭제
 (bugpigg/issue-tracker#45)

mybloom committed:
feat: IssueAssignee  유효성 체크 로직을 IssueAssigneeService로 이동
 (bugpigg/issue-tracker#45)

mybloom committed:
fix: 컬렉션 하나씩 delete하지 않고 deleteAll()로 한번에 삭제
 (bugpigg/issue-tracker#45)

mybloom committed:
fix: 잘못 작성된 service 삭제
 (bugpigg/issue-tracker#45)

mybloom committed:
Merge branch 'BE-feat-45' into BE-5

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/IssueRepository.java:
```diff
@@ -6,11 +6,38 @@
+		+ "AND (coalesce(:milestones, NULL) IS NULL OR M.id IN (:milestones)) "
+		+ "AND (:title IS NULL OR I.title LIKE :title)"
+	)
+	List<Issue> findBySearchCondition(@Param("opened") boolean opened,
```
이런 부분은 쿼리 DSL 적용을 고려해보면 좋을 것 같습니다!

bugpigg replied:
넵 쿼리 DSL도 추후에 공부해봐야겠습니다!

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/IssueRepository.java:
```diff
@@ -6,11 +6,38 @@
+		+ "AND (coalesce(:writers, NULL) IS NULL OR I.memberId IN (:writers)) "
+		+ "AND (coalesce(:labels, NULL) IS NULL OR L.id IN (:labels)) "
+		+ "AND (coalesce(:milestones, NULL) IS NULL OR M.id IN (:milestones)) "
+		+ "AND (:title IS NULL OR I.title LIKE :title)"
```
완전히 일치하는 경우만 조회하나요?

bugpigg replied:
DTO 에서 와일드카드를 추가해주고 있습니다!

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/IssueService.java:
```diff
@@ -309,4 +310,24 @@ private void validateLabelIds(List<Long> labelIds, Long memberId) {
+	@Transactional(readOnly = true)
+	public List<IssueListResponseDto> findBySearchCondition(IssueSearchRequestDto searchRequestDto,
+		Long currentMemberId) {
+		searchRequestDto.setCurrentMemberId(currentMemberId);
+		searchRequestDto.addCurrentMemberToWriters(currentMemberId);
```
밑에서 바로 꺼내는 것 같은데 넣어주는 이유가 있나요?

Dae-Hwa replied:
그리고 작성자가 내가 아닌 경우는 조회할 수 없게되는걸까요?

bugpigg replied:
설정할 수 있는 검색 조건이 `내가 쓴 이슈` 와 `작성자` 가 있습니다.
이 두 조건은 중복 설정이 가능하다고 생각하여, 하나의 `writer` 리스트에서 관리하게 하였습니다.

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/dto/IssueSearchRequestDto.java:
```diff
@@ -0,0 +1,38 @@
+	private List<Long> milestoneId;
+
+	public void addCurrentMemberToWriters(Long currentMemberId) {
+		if (writtenByMe){
+			if (writerId == null) {
+				writerId = new ArrayList<>();
+			}
+			writerId.add(currentMemberId);
+		}
```
DTO는 데이터 전달을 위한건데, 이런 부분까지 책임져도 괜찮을까요?

bugpigg replied:
DTO 라는 이름을 가진 클래스에 기대되지 않는 부분인것 같습니다
서비스레이어에서 처리하도록 수정하는게 좋을것 같습니다!

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/dto/IssueSearchRequestDto.java:
```diff
@@ -0,0 +1,38 @@
+
+	public void trimAndFormattingTitle() {
+		if (title != null) {
+			title = "%" + title.trim() + "%";
```
여기서 와일드카드 처리를 해주는군요
trim은 DTO에 넣으면서 해줄 수도 있고, 서비스에서 꺼내면서 해도 괜찮을 것 같아요
% 와일드카드는 쿼리에 포함시켜줘야 어떤 동작을 하는 쿼리인지 알기 쉬울 것 같아요

bugpigg replied:
오 넵 쿼리에서 직접적으로 표현하는 방식이 다른 분들이 이해하기 쉬운 코드가 될것 같습니다!
수정하겠습니다!

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/IssueAssigneeService.java:
```diff
@@ -0,0 +1,47 @@
+				issueAssigneeRepository.saveAll(issueAssignees));
+
+			//연관관계 편의메서드
+			issue.addIssueAssignee(assigneeSet);
```
IssueAssignee 를 생성하면서 하나씩 넣을 수도 있지 않았을까요?

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/IssueController.java:
```diff
@@ -68,6 +74,13 @@ public ResponseEntity<SelectableLabelMilestoneResponse> readyToEditLabelAndMiles
 	public ResponseEntity<CommonResponseDto> create(
 		@RequestBody IssueSaveRequestDto issueSaveRequestDto) {
+
+		milestoneService.validateMilestoneId(issueSaveRequestDto.getMilestoneId(), MEMBER_ID);
+
+		labelService.validateLabelIds(issueSaveRequestDto.getLabelIds(), MEMBER_ID);
+
+		issueAssigneeService.validateAssigneeIds(issueSaveRequestDto.getAssigneeIds());
+
```
컨트롤러에서 검증해야하는건 요청값이 유효하느냐(null이 아닌지, 비어있지는 않은지, 숫자가 와야하는데 문자열이 왔다던지 등)지 내부에 저장되어있는지가 아닌 것 같아요

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/IssueLabelService.java:
```diff
@@ -0,0 +1,31 @@
+				.map(labelId -> IssueLabel.of(issue.getId(), labelId))
+				.collect(Collectors.toList());
+
+			Set<IssueLabel> savedIssueLabel = Set.copyOf(issueLabelRepository.saveAll(issueLabels));
```
복사해준 이유가 있나요?

Dae-Hwa replied:
이 메소드만 놓고 봤을때 저장에 실패한 label이 있어도 롤백되거나 하지 않아도 상관없나요?

mybloom replied:
리뷰 내용에 힌트를 얻어 save()와 saveAll()의 차이를 알아보았는데요. 
saveAll()은 배치insert가 이뤄지는 작업으로 JPA에 별도의 셋팅을 해줘야 하고
GenerationType.IDENTITY에서는 이뤄지지 않는다는 것을 알았습니다. 

각각의 엔티티를 save()로 저장하면서 연관관계 메서드도 설정하는 방법으로 변경하겠습니다.

그리고 해당 메서드에 `@Transactional`을 명시하지 않았는데요. 
이 메서드도 `이슈 저장` 시 호출되는 메서드의 일부라서 최초 메서드에 트랜잭션이 선언되어 있습니다

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/issue/domain/Issue.java:
```diff
@@ -50,20 +53,24 @@ public class Issue extends BaseTimeEntity {
 	}
 
-	public void addIssueLabel(List<IssueLabel> issueLabels) {
+	public void addIssueLabel(Set<IssueLabel> issueLabels) {
```
add면 add를 해줘야 하는거 아닐까요? `this.issueLabels = issueLabels` 와 같이 할당해주면 이전 label이 계속 있다고 할 수 있나요?

Dae-Hwa commented on BE/src/main/java/com/team09/issue_tracker/label/LabelService.java:
```diff
@@ -10,9 +11,20 @@ public class LabelService {
+		}
+	}
 
+	private void validateMyLabelIds(List<Long> labelIds, Long memberId) {
+		if (!isMyLabels(labelIds, memberId)) {
+			throw new LabelInvalidException();
+		}
+	}
```
제가 봤을때는 특별한 동작 없이 validateLabelIds와 isMyLabels을 연결만 시켜주는 메소드인 것 같은데 어떤 것 같으세요?

mybloom replied:
네 불필요한 메서드 인 것 같습니다~

mybloom replied:
네 불필요한 메서드 같습니다. 바로 isMyLabels()를 사용해도 되는 것 같습니다.

Dae-Hwa reviewed (commented):
테리, 벅픽 안녕하세요!

짧은 시간인데 읽기가 한결 편해진 것 같네요

고생많으셨습니다~ㅎㅎ

mybloom commented:
안녕하세요 프레디~
오늘은 날씨가 쨍쨍하네요!

어제는 수료식이 있었는데요, 
오늘 리뷰에 커멘트 달아보도록 하겠습니다.
리뷰 감사합니다!

bugpigg committed:
[BE] 이슈 검색 관련 수정 사항 반영 (#53)

* fix: 이슈 조건 검색 쿼리에 와일드 카드를 나타내도록 수정

- 이슈 조건 검색 쿼리에 와일드 카드를 나타내도록 수정하였습니다.
- 그리고 스트링의 트림 함수는 DTO의 필드 값이 선언될때 적용하도록 수정하였습니다.

* fix: 작성자 리스트에 현재 유저 아이디 추가 메소드 수정

- 기존의 DTO에 위치하던 메소드를 서비스 레이어에 위치하도록 수정하였습니다.
